Add initial documentation

hsorby opened the pull request:
on setup, building, coding standard, contribution and review.  Some images are missing, coming soon.

Addresses issue #13

hsorby committed:
Add some early documentation on setup, building, coding standard, contribution and review.  Some images are missing, coming soon.

hsorby committed:
Add documentation describing contribution, also added a glossary.

hsorby committed:
Add images.  Add documentation on how to get code tested via Buildbot before creating a pull request.

hsorby committed:
Adding git branching image

Commit:
update copyright statement

Commit:
formatting tweaks; use numref to link to labelled figures

Commit:
slight tweak to the grammar

hsorby committed:
Merge pull request #1 from nickerso/basic-docs

Slight improvements, thanks for reading it.

nickerso commented:
I think this is a good start, and certainly sufficient to get my own libCellML fork and clone up and building. The documentation will obviously need work to complete it over time, but I am happy with it meeting the needs of issue #13.

Also a big thanks to @hsorby for the cool images!

jonc125 commented on README.rst:
```diff
@@ -7,3 +7,8 @@ The prime libCellML repository is located at: https://github.com/cellml/libcellm
+Scope
+=====
+
+The purpose of libCellML is to create, manipulate, serialize, deserialize, validate and instantiate CellML models.  It is not concerned with retrieving CellML imports from external sources.
```
I'm pretty sure some of the initial use cases concerned dealing with imports? It seems to me that resolving external imports is the main headache in going from CellML 1.0 to 1.1, so if the library doesn't do this I don't gain much beyond what I already have!

nickerso replied:
Good point Jonathan. I think we should discuss this a bit more on the mailing list before adding it in. @hsorby, perhaps remove this change from this set and I'll create a new issue regarding the scope.

jonc125 commented on docs/conf.py:
```diff
@@ -48,16 +48,16 @@
 # General information about the project.
 project = u'libCellML'
-copyright = u'2014, David Nickerson'
+copyright = u'2014-2015, University of Auckland'
```
This could be an issue for external contributors. (All my work is officially owned by Oxford.)

Not that we've managed to get research services here to sort out a proper contributor licence agreement for Chaste, having been asking for >5 years, but you may have more luck at Auckland!

nickerso replied:
yeah, I just made this change to bring the docs in line with the current copyright statement in the rest of the code - its not a fixed thing that is required from this end. Is there a standard way to say "copyright the contributors" ? or something similar? maybe a "Copyright libCellML Contributors" and then list contributors somewhere?

agarny replied:
I have done something similar to "Copyright libCellML Contributors" for OpenCOR. See [here](https://github.com/opencor/opencor/blob/master/NOTICE.txt).

agarny replied:
Just noticed that there are a few other places where we might want to remove David's name. Lines 204-207:

```
latex_documents = [
  ('index', 'libCellML.tex', u'libCellML Documentation',
   u'David Nickerson', 'manual'),
]
```

Lines 234-237:

```
man_pages = [
    ('index', 'libcellml', u'libCellML Documentation',
     [u'David Nickerson'], 1)
]
```

And lines 248-252:

```
texinfo_documents = [
  ('index', 'libCellML', u'libCellML Documentation',
   u'David Nickerson', 'libCellML', 'One line description of project.',
   'Miscellaneous'),
]
```

jonc125 commented on docs/dev_building_from_source.rst:
```diff
@@ -0,0 +1,47 @@
+Building libCellML from Source
+==============================
+
+This document covers building libCellML from source.  It is assumed that you already have the codebase downloaded and ready for building.  The variable LIBCELLML_SRC shall be used to refer to the directory containing the LICENSE file for libCellML.
```
The naming here should be updated to match the camel case from #10.

nickerso replied:
Do you think the name of the project needs to match the camel case used in the code? The name of the project is libCellML, but that doesn't work so well in actual code (hence LibCellml (upper-) or libCellml (lower camel case)). If its better they match, then I'd change my vote in the camel case discussion ;-)

agarny replied:
Agreed with @nickerso. If at all possible, I would keep libCellML for the project name, but LibCellml in the code. If not, then I would also change my vote.

jonc125 replied:
Hmm, I see your points. libCellML does work better as a project name, but I don't like creating an exception to the naming convention in code. So happy to go with the status quo.

jonc125 commented on docs/dev_building_from_source.rst:
```diff
@@ -0,0 +1,47 @@
+
+.. include:: dev_configuration_options.rst
+
+.. Select a generator that is appropriate for your system, or let the CMake configuration application 
```
Something's missing here :)

jonc125 commented on docs/dev_coding_standard.rst:
```diff
@@ -0,0 +1,13 @@
+
+The coding standard for libCellML follows that specified in the `google c++ style guide <http://google-styleguide.googlecode.com/svn/trunk/cppguide.html>`_.  But it does not have to be followed to the letter, `other people <https://www.linkedin.com/pulse/20140503193653-3046051-why-google-style-guide-for-c-is-a-deal-breaker>`_ have not so positive opinions about the quality of the google style gui ...
+
+In essence we seek code that looks good, is easy to read and has great documentation.  We don't want to spend time discussing the minutiae of the coding style.
```
Good!

Perhaps add a comment about try to be consistent with the style of existing code you add to?

agarny replied:
+1 regarding @jonc125's suggestion.

jonc125 commented on docs/glossary.rst:
```diff
@@ -0,0 +1,13 @@
+   :sorted:
+
+   Prime repository
+   Prime libCellML repository
```
Naming again

jonc125 commented:
I've not followed the steps to check I can fork & build, but it all reads well and looks sensible to me.

hsorby committed:
Remove premature scope text.

hsorby committed:
Merge branch 'basic-docs' of github.com:hsorby/libcellml into basic-docs

agarny commented on docs/dev_building.rst:
```diff
@@ -0,0 +1,12 @@
+Building from Source
+====================
+
+See :doc:`Building from Source <dev_building_from_source>` on how to build libCellML from source
```
Very minor point: might want to add a newline at the end of the file.

agarny commented on docs/dev_building_from_source.rst:
```diff
@@ -0,0 +1,47 @@
+
+The libCellML library uses the CMake build configuration tool to configure the library.  Version 3.2 or greater of CMake is required to configure libCellML.  Download and installation instructions are available from `CMake <http://www.cmake.org/>`_.
+
+The configuration options for the library are detailed in the following table.  The command line options can be set with the -D flag, like so -DBUILD_TYPE=Release.  Please note that in CMake GUI Configuration applications the config variable is prefixed with 'LIBCELLML\_'
```
To make it easier to read, we might want to use ```` to render things like `-DBUILD_TYPE=Release`, i.e. use ```-DBUILD_TYPE=Release``` (and the same wherever else it would be useful).

agarny commented on docs/dev_configuration_options.rst:
```diff
@@ -0,0 +1,45 @@
+-------------- ------------ -----------------------------------------
+COVERAGE       ON           Enable coverage testing. (if available)
+============== ============ =========================================
+
```
Minor stuff: a full stop is used for some descriptions and not for some others. Otherwise, another case where `BUILD_TYPE` and `Release` (and the like) could be written using ```BUILD_TYPE``` and ```Release```.

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+:numref:`Figure %s <libcellml-development-process>` gives a graphical overview of the developer contribution process.  For more detail see the text below.
+
+.. _libcellml-development-process:
+.. figure:: images/libCellMLProcesses-DevelopmentProcess.png
```
A couple of very minor points:
- You might want to replace **cellml/libcellml** with **CellML/libCellML**.
- Missing full point at the end of **Reviewer to merge pull request**.

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+Topic Branch
+============
+
+A topic or feature branch is a branch that is local to you (and anyone you collaborate with), it is branch that will not be available from the :term:`prime repository`.  All development work should be carried out on a topic branch, for example any major feature that you work on or minor bug fix.  Before creating a local topic branch pull the latest changes from the :term:`prime repository` develo ...
```
**it is [a] branch that...**

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+A topic or feature branch is a branch that is local to you (and anyone you collaborate with), it is branch that will not be available from the :term:`prime repository`.  All development work should be carried out on a topic branch, for example any major feature that you work on or minor bug fix.  Before creating a local topic branch pull the latest changes from the :term:`prime repository` develo ...
+
+.. _libcellml-branching:
+.. figure:: images/libCellMLProcesses-GitBranching.png
```
Shouldn't the **pull/fetch** dashed arrow between the **CellML** and **PC** 'boxes' be going from **CellML** to **PC** and not the other way round?

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+   git fetch prime develop
+   git checkout develop # Not required if already on develop branch
+   git merge prime/develop
+   git checkout -b iss46
```
Very minor: add two characters and make it clearer/obvious: `issue46` rather than `iss46`.

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+GitHub Pull Request
+===================
+
+Once some changes have been made and local commits committed push your changes to your GitHub libcellml repository (refer to :numref:`Figure %s <libcellml-branching>`).  From there create a pull request from your topic branch to the :term:`prime repository` develop branch.  When creating the pull request make sure to add in the comment 'addresses issue #46' (of course replace the number 46 with t ...
```
libCellML, not libcellml (at least, that's what is done below and what I would do too).

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+   git commit -m "Appropiately descriptive message about the changes made."
+   git push -u origin iss46
+
+The 'git add' and 'git commit' commands should be obvious,  the 'git push' command sets the local branch 'iss46' to be linked with the remote branch 'iss46' in the origin (the default shorthand for your libCellML repository on GitHub) repository, this branch will be created in the origin repository if it doesn't already exist.  
```
Replace `'git add'` with ```git add```, etc.

agarny commented on docs/dev_contribution.rst:
```diff
@@ -0,0 +1,95 @@
+Completion
+==========
+
+Once the owners of the :term:`prime repository` are satisfied that the work on the issue is complete and that the feedback has been addressed, one of the owners will post a comment on the issue notifying subscribers that they intend to merge the pull request.  If no further objections are raised the owner will merge an close the pull request as outlined in :doc:`Reveiw Process <dev_review>`.
```
If I recall correctly, one of the owners must say that s/he is happy with the pull request. Then, a second owner must also say that s/he is happy. Once that is done, and because we have two 'happy' owners, the second owner can merge the pull request.

Minor point: `:doc:`Review Process <dev_review>`` and not `:doc:`Reveiw Process <dev_review>``.

agarny commented on docs/dev_setup.rst:
```diff
@@ -0,0 +1,88 @@
+
+The libCellML codebase is hosted on GitHub and therefore Git is used to track changes in the codebase.  Before you begin you will need to have a few prerequisites satisfied:
+
+#. Github user account (for the rest of this document we will call our user *andre*)
```
Minor correction: GitHub and not Github.

agarny commented on docs/dev_setup.rst:
```diff
@@ -0,0 +1,88 @@
+
+.. _fig_devSetup_githubRepos:
+
+.. figure:: images/libCellMLProcesses-GitHubRepos.png
```
Shouldn't the **prime (read only)** dashed arrow between the **CellML** and **PC** 'boxes' be going from **CellML** to **PC** and not the other way round?

agarny commented on docs/dev_setup.rst:
```diff
@@ -0,0 +1,88 @@
+        cd <somewhere/you/keep/development/code>
+        git clone git@github.com:andre/libcellml.git
+
+Note: Again, don't and clone this location substitute your username for *andre*.
```
**don't and clone**?...

Otherwise, I find the text in general to be missing a few commas here and there.

hsorby committed:
Merge branch 'develop' into basic-docs

hsorby committed:
Address comments from pull request #14.

hsorby committed:
Fix minor issues, crop back object model images, remove \ character.

hsorby commented:
I have addressed all the comments given here in the recent commits.  The only thing I haven't done is mark up the variables in the table of configuration options, that just looked really weird.  I figured that a table is a table and it doesn't require the markup.  If more than one person disagrees with this rationalisation I will change it.

agarny commented:
Apart from my minor comment above, this all looks good to me.

hsorby committed:
Reword a little of the text in Completion section of dev_contribution and the Coding Standard section of dev_review.

nickerso commented:
I think this is now all good and addresses all the comments made.

hsorby committed:
Correct bad england in dev_contribution.
